### PR TITLE
Add reconstructed electron factory, algorithm utilizing E/p cut 

### PR DESCRIPTION
**Pull request by jdbrice**
### Briefly, what does this PR introduce?
Provides electron candidates as a first step towards a DIS electron finder.
This PR uses the recently provided MC particle to Reco cluster associations to produce a `ReconstructedElectrons` collection. Electrons candidates are identified with an E/p cut using Reco particle momentum and Redo cluster energy (only from ECAL detectors for now).

Couple of comments/potential improvements for reviewers:
- I am not sure how best to use the Config objects to allow parameters to be modified (e.g. min/max E/p in this case)
- AFAIK this creates a copy of the reconstructed particles - since this (ReconstructedElectrons) is a strict subset of the ReconstructedParticles collection, is there a way to store as such (without duplicating)
- I am currently pulling from (by name) several locations (see `reco.cc`) to get all possible source of `edm4eic::MCRecoClusterParticleAssociation` - is there a way to instead ask for all sources of a given type?

### What kind of change does this PR introduce?
- [ ] Bug fix (issue #__)
- [x] New feature (issue #__)
- [ ] Documentation update
- [ ] Other: __

### Please check if this PR fulfills the following:
- [ ] Tests for the changes have been added: 
  - Not sure if new tests needed, 
  - "ReconstructedElectrons" added to polio default output (include list) - simply running on a recent sim file should suffice as a test - expected behavior is to produce the "ReconstructedElectrpons" collection. A second level of test is to ensure that the collection is not empty (assuming at least some electrons within acceptance are in the simulated input)
- [ ] Documentation has been added / updated
- [x] Changes have been communicated to collaborators: discussion in S&C meeting on June 5

### Does this PR introduce breaking changes? What changes might users need to make to their code?
No

### Does this PR change default behavior?
No

### CODE REVIEW

File: src/global/reco/ReconstructedElectrons_factory.h [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        // Step 3. Pass everything to "the algorithm"
+        // in the future, select appropriate algorithm (truth, fully reco, etc.)
+        auto output = m_algo.execute(
+          mc_particles,
+          rc_particles,
+          rc_particles_assoc,
+          in_clu_assoc
+        );
+
+        m_log->info( "We have found {} reconstructed electron candidates this event", output.size() );
```
Comment (veprbl) [inline src/global/reco/ReconstructedElectrons_factory.h:86]: Let's not spam logs
```suggestion
        m_log->debug( "We have found {} reconstructed electron candidates this event", output.size() );
```

File: src/global/reco/ReconstructedElectrons_factory.h [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        m_log->debug("ReconstructedElectrons_factory: plugin='{}' prefix='{}'", plugin, prefix);
+      }
+
+      /** On run change preparations **/
+      void BeginRun(const std::shared_ptr<const JEvent> &event) override{}
+
+      /** Event by event processing **/
+      void Process(const std::shared_ptr<const JEvent> &event) override{
+        m_log->debug("ReconstructedElectrons_factory: Process");
+
```
Comment (veprbl) [inline src/global/reco/ReconstructedElectrons_factory.h:57]:
```suggestion
```

File: src/algorithms/reco/ElectronReconstruction.cc [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        // Step 2. Get Reco particle for the Mc Particle matched to cluster
+        // Step 3. Apply E/p cut using Reco cluster Energy and Reco Particle momentum
+
+        // Some obvious improvements:
+        // - E/p cut from real study optimized for electron finding and hadron rejection
+        // - use of any HCAL info?
+        // - check for duplicates?
+
+        m_log->debug("ElectronReconstruction::execute");
+
```
Comment (veprbl) [inline src/algorithms/reco/ElectronReconstruction.cc:40]:
```suggestion
```

File: src/global/reco/reco.cc [hunk 4]
```diff
@@ -70,5 +71,18 @@ void InitPlugin(JApplication *app) {
     app->Add(new JChainFactoryGeneratorT<ReconstructedJets_factory>(
             {"ReconstructedParticles"}, "ReconstructedJets"));
 
+    app->Add(new JChainFactoryGeneratorT<ReconstructedElectrons_factory>(
+        {"MCParticles", "ReconstructedChargedParticles", "ReconstructedChargedParticleAssociations",
+        "EcalBarrelScFiClusterAssociations",
+        // "EcalBarrelSciGlassClusterAssociations",
```
Comment (veprbl) [inline src/global/reco/reco.cc:77]: Ignore SciGlass
```suggestion
```

File: src/algorithms/reco/ElectronReconstruction.cc [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+                found_reco_part = true;
+                break;
+              }
+            }
+
+            // if we found a reco particle then test for electron compatibility
+            if ( found_reco_part ){
+              auto reco_part = (*reco_part_assoc)->getRec();
+              double EoverP = clu.getEnergy() / edm4eic::magnitude(reco_part.getMomentum());
+              m_log->info( "ReconstructedParticle: Energy={}, p={}, E/p = {} for PDG (from truth): {}", clu.getEnergy(), edm4eic::magnitude(reco_part.getMomentum()), EoverP, sim.getPDG() );
```
Comment (veprbl) [inline src/algorithms/reco/ElectronReconstruction.cc:68]: Logs from inside loops should be even below debug.
```suggestion
              m_log->trace( "ReconstructedParticle: Energy={}, p={}, E/p = {} for PDG (from truth): {}", clu.getEnergy(), edm4eic::magnitude(reco_part.getMomentum()), EoverP, sim.getPDG() );
```

File: src/global/reco/ReconstructedElectrons_factory.h [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+        // get plugin name and tag
+        auto app    = GetApplication();
+        auto plugin = GetPluginName();
+        auto prefix = plugin + ":" + GetTag();
+        InitDataTags(prefix);
+
+        // services
+        InitLogger(prefix, "info");
+        m_algo.init(m_log);
+        m_log->debug("ReconstructedElectrons_factory: plugin='{}' prefix='{}'", plugin, prefix);
```
Comment (veprbl) [inline src/global/reco/ReconstructedElectrons_factory.h:48]:
```suggestion
```

File: src/global/reco/ReconstructedElectrons_factory.h [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+// services
+#include <services/log/Log_service.h>
+#include <extensions/spdlog/SpdlogExtensions.h>
+#include <extensions/spdlog/SpdlogMixin.h>
+
+namespace eicrecon {
+
+// algorithm
+  class ElectronReconstruction;
+
```
Comment (veprbl) [inline src/global/reco/ReconstructedElectrons_factory.h:25]: Forward declaration not needed. You embed the algorithm as non-pointer member, so header inclusion is required.
```suggestion
```

File: src/global/reco/ReconstructedElectrons_factory.h [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+#pragma once
+
+// JANA
+#include <extensions/jana/JChainFactoryT.h>
+#include <JANA/JEvent.h>
+
+// data model
+#include <edm4eic/TrackSegmentCollection.h>
+
```
Comment (veprbl) [inline src/global/reco/ReconstructedElectrons_factory.h:12]:
```suggestion
```
Unused

File: src/algorithms/reco/ElectronReconstruction.cc [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+        // output container
+        std::vector<edm4eic::ReconstructedParticle*> electrons_edm;
+
+        for ( auto col : in_clu_assoc ){ // loop on cluster association collections
+          for ( auto clu_assoc : col ){ // loop on MCRecoClusterParticleAssociation in this particular collection
+            auto sim = clu_assoc->getSim(); // McParticle
+            auto clu = clu_assoc->getRec(); // RecoCluster
+
+            m_log->debug( "SimId={}, CluId={}", clu_assoc->getSimID(), clu_assoc->getRecID() );
+            m_log->debug( "MCParticle: Energy={}, p={}, E/p = {} for PDG: {}", clu.getEnergy(), edm4eic::magnitude(sim.getMomentum()), clu.getEnergy() / edm4eic::magnitude(sim.getMomentum()), sim.getPDG() );
```
Comment (veprbl) [inline src/algorithms/reco/ElectronReconstruction.cc:50]:
```suggestion
            m_log->trace( "SimId={}, CluId={}", clu_assoc->getSimID(), clu_assoc->getRecID() );
            m_log->trace( "MCParticle: Energy={}, p={}, E/p = {} for PDG: {}", clu.getEnergy(), edm4eic::magnitude(sim.getMomentum()), clu.getEnergy() / edm4eic::magnitude(sim.getMomentum()), sim.getPDG() );
```

File: src/algorithms/reco/ElectronReconstruction.cc [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+
+  void ElectronReconstruction::init(std::shared_ptr<spdlog::logger> logger) {
+    m_log = logger;
+  }
+
+  std::vector<edm4eic::ReconstructedParticle*> ElectronReconstruction::execute(
+    std::vector<const edm4hep::MCParticle *> mcparts,
+    std::vector<const edm4eic::ReconstructedParticle *> rcparts,
+    std::vector<const edm4eic::MCRecoParticleAssociation *> rcassoc,
+    std::vector<std::vector<const edm4eic::MCRecoClusterParticleAssociation*>> &in_clu_assoc
```
Comment (veprbl) [inline src/algorithms/reco/ElectronReconstruction.cc:27]:
```suggestion
    const std::vector<const edm4hep::MCParticle *> &mcparts,
    const std::vector<const edm4eic::ReconstructedParticle *> &rcparts,
    const std::vector<const edm4eic::MCRecoParticleAssociation *> &rcassoc,
    const std::vector<std::vector<const edm4eic::MCRecoClusterParticleAssociation*>> &in_clu_assoc
```

File: src/algorithms/reco/ElectronReconstruction.cc [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+#include "edm4eic/MCRecoClusterParticleAssociationCollection.h"
+
+namespace eicrecon {
+
+  void ElectronReconstruction::init(std::shared_ptr<spdlog::logger> logger) {
+    m_log = logger;
+  }
+
+  std::unique_ptr<edm4eic::ReconstructedParticleCollection> ElectronReconstruction::execute(
+    const std::vector<const edm4hep::MCParticle *> &mcparts,
```
Comment (wdconinc) [inline src/algorithms/reco/ElectronReconstruction.cc:25]: Why not `MCParticleCollection`?

File: src/algorithms/reco/ElectronReconstruction.cc [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+    const std::vector<const edm4eic::ReconstructedParticle *> &rcparts,
```
Comment (wdconinc) [inline src/algorithms/reco/ElectronReconstruction.cc:26]: Why no `ReconstructedParticleCollection`?

File: src/algorithms/reco/ElectronReconstruction.cc [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+    const std::vector<const edm4eic::MCRecoParticleAssociation *> &rcassoc,
```
Comment (wdconinc) [inline src/algorithms/reco/ElectronReconstruction.cc:27]: Why no `MCRecoParticleAssociationCollection`?

File: src/algorithms/reco/ElectronReconstruction.cc [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+        // Some obvious improvements:
+        // - E/p cut from real study optimized for electron finding and hadron rejection
+        // - use of any HCAL info?
+        // - check for duplicates?
+
+        // output container
+        auto out_electrons = std::make_unique<edm4eic::ReconstructedParticleCollection>();
+
+        for ( auto col : in_clu_assoc ){ // loop on cluster association collections
+          for ( size_t i = 0; i < col->size(); i++ ){ // loop on MCRecoClusterParticleAssociation in this particular collection
```
Comment (wdconinc) [inline src/algorithms/reco/ElectronReconstruction.cc:44]: Use ranged for.

File: src/algorithms/reco/ElectronReconstruction.cc [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+            auto clu = clu_assoc.getRec(); // RecoCluster
+
+            m_log->trace( "SimId={}, CluId={}", clu_assoc.getSimID(), clu_assoc.getRecID() );
+            m_log->trace( "MCParticle: Energy={} GeV, p={} GeV, E/p = {} for PDG: {}", clu.getEnergy(), edm4eic::magnitude(sim.getMomentum()), clu.getEnergy() / edm4eic::magnitude(sim.getMomentum()), sim.getPDG() );
+
+
+            // Find the Reconstructed particle associated to the MC Particle that is matched with this reco cluster
+            // i.e. take (MC Particle <-> RC Cluster) + ( MC Particle <-> RC Particle ) = ( RC Particle <-> RC Cluster )
+            auto reco_part_assoc = rcassoc.begin();
+            bool found_reco_part = false;
```
Comment (wdconinc) [inline src/algorithms/reco/ElectronReconstruction.cc:56]: Prefer to loop without sentinel and use equality with end as marker.

File: src/algorithms/reco/ElectronReconstruction.h [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+
+#include <spdlog/spdlog.h>
+
+#include <edm4eic/vector_utils.h>
+
+// Event Model related classes
+#include "edm4hep/MCParticleCollection.h"
+#include "edm4eic/ReconstructedParticleCollection.h"
+#include "edm4eic/MCRecoParticleAssociationCollection.h"
+#include "edm4eic/MCRecoClusterParticleAssociation.h"
```
Comment (wdconinc) [inline src/algorithms/reco/ElectronReconstruction.h:19]: Collection header includes type itself so this is redundant with the next include.

File: src/global/reco/ReconstructedElectrons_factory.h [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+        // Step 1. lets collect the Cluster associations from various detectors
+        std::vector<const edm4eic::MCRecoClusterParticleAssociationCollection*> in_clu_assoc;
+        for(auto& input_tag : GetInputTags()){
+          // only collect from the sources that provide ClusterAssociations
+          if ( input_tag.find( "ClusterAssociations" ) == std::string::npos ) {
+            continue;
+          }
+          m_log->trace( "Adding cluster associations from: {}", input_tag );
+          in_clu_assoc.push_back(
+            static_cast<const edm4eic::MCRecoClusterParticleAssociationCollection*>(event->GetCollectionBase(input_tag))
```
Comment (wdconinc) [inline src/global/reco/ReconstructedElectrons_factory.h:58]: Is this used elsewhere? I don't recall the need for static_cast.

File: src/global/reco/ReconstructedElectrons_factory.h [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+          );
+        }
+
+        // Step 2. Get MC, RC, and MC-RC association info
+        // This is needed as a bridge to get RecoCluster - RC Particle associations
+        auto mc_particles = event->Get<edm4hep::MCParticle>("MCParticles");
```
Comment (wdconinc) [inline src/global/reco/ReconstructedElectrons_factory.h:64]: GetCollection.

File: src/global/reco/ReconstructedElectrons_factory.h [hunk 19]
```diff
@@ -0,0 +1,88 @@
...
+        auto rc_particles = event->Get<edm4eic::ReconstructedParticle>("ReconstructedChargedParticles");
```
Comment (wdconinc) [inline src/global/reco/ReconstructedElectrons_factory.h:65]: Id

File: src/global/reco/ReconstructedElectrons_factory.h [hunk 20]
```diff
@@ -0,0 +1,88 @@
...
+        auto rc_particles_assoc = event->Get<edm4eic::MCRecoParticleAssociation>("ReconstructedChargedParticleAssociations");
```
Comment (wdconinc) [inline src/global/reco/ReconstructedElectrons_factory.h:66]: Id

**Review by wdconinc — COMMENTED** [review]
Podio supports reference/subset collections that only refer to a subset of another collection without making a copy. The canonical example of a use case for that is exactly this: selection a dis electron. I don't think we use it yet, and I don't know if jana2/EICrecon supports it. @nathanwbrei may know.

**Comment by veprbl** [comment]
See example of conversion to passing collections in #741

File: src/global/reco/ReconstructedElectrons_factory.h (hunk 17 above)
Comment (jdbrice) [reply src/global/reco/ReconstructedElectrons_factory.h:58]: yes, see mattermost discussion, GetCollection<> is giving problems and this was suggested as a workaround

File: src/global/reco/ReconstructedElectrons_factory.h (hunk 20 above)
Comment (jdbrice) [reply src/global/reco/ReconstructedElectrons_factory.h]: done with bf659cb

File: src/global/reco/ReconstructedElectrons_factory.h (hunk 19 above)
Comment (jdbrice) [reply src/global/reco/ReconstructedElectrons_factory.h]: done with bf659cb

File: src/global/reco/ReconstructedElectrons_factory.h (hunk 18 above)
Comment (jdbrice) [reply src/global/reco/ReconstructedElectrons_factory.h]: See above, done with bf659cb

File: src/algorithms/reco/ElectronReconstruction.h (hunk 16 above)
Comment (jdbrice) [reply src/algorithms/reco/ElectronReconstruction.h]: done with bf659cb

File: src/algorithms/reco/ElectronReconstruction.cc (hunk 15 above)
Comment (jdbrice) [reply src/algorithms/reco/ElectronReconstruction.cc]: done with bf659cb

File: src/algorithms/reco/ElectronReconstruction.cc (hunk 14 above)
Comment (jdbrice) [reply src/algorithms/reco/ElectronReconstruction.cc]: done with bf659cb

File: src/algorithms/reco/ElectronReconstruction.cc (hunk 13 above)
Comment (jdbrice) [reply src/algorithms/reco/ElectronReconstruction.cc]: done with 7dab36f

File: src/algorithms/reco/ElectronReconstruction.cc (hunk 12 above)
Comment (jdbrice) [reply src/algorithms/reco/ElectronReconstruction.cc]: done with 7dab36f

File: src/algorithms/reco/ElectronReconstruction.cc (hunk 11 above)
Comment (jdbrice) [reply src/algorithms/reco/ElectronReconstruction.cc]: done with 7dab36f

File: src/algorithms/reco/ElectronReconstruction.cc [hunk 21]
```diff
@@ -0,0 +1,83 @@
+// SPDX-License-Identifier: LGPL-3.0-or-later
+// Copyright (C) 2023 Daniel Brandenburg
+
+#include <algorithm>
+#include <cmath>
+#include <vector>
+
+#include "Beam.h"
+#include "Boost.h"
```
Comment (wdconinc) [inline src/algorithms/reco/ElectronReconstruction.cc:9]: Unused, I think.

Comment (wdconinc) [reply src/algorithms/reco/ElectronReconstruction.cc:9]: Probably `cmath` is unused too.

File: src/algorithms/reco/ElectronReconstruction.h [hunk 22]
```diff
@@ -0,0 +1,44 @@
+// SPDX-License-Identifier: LGPL-3.0-or-later
+// Copyright (C) 2023 Daniel Brandenburg
+
+#pragma once
+
+#include <algorithm>
+#include <cmath>
```
Comment (wdconinc) [inline src/algorithms/reco/ElectronReconstruction.h:7]: Probably unused.

File: src/algorithms/reco/ElectronReconstruction.h [hunk 23]
```diff
@@ -0,0 +1,44 @@
...
+#include <vector>
+#include <map>
```
Comment (wdconinc) [inline src/algorithms/reco/ElectronReconstruction.h:9]: Unused.

File: src/algorithms/reco/ElectronReconstruction.h [hunk 24]
```diff
@@ -0,0 +1,44 @@
+// SPDX-License-Identifier: LGPL-3.0-or-later
+// Copyright (C) 2023 Daniel Brandenburg
+
+#pragma once
+
+#include <algorithm>
```
Comment (wdconinc) [inline src/algorithms/reco/ElectronReconstruction.h:6]: Probably unused in header at least.

File: src/algorithms/reco/ElectronReconstruction.h [hunk 25]
```diff
@@ -0,0 +1,44 @@
...
+#include <cmath>
+#include <vector>
+#include <map>
+
+#include <spdlog/spdlog.h>
+
+#include <edm4eic/vector_utils.h>
```
Comment (wdconinc) [inline src/algorithms/reco/ElectronReconstruction.h:13]: Probably also unused.

**Review by wdconinc — APPROVED** [review]
Some minor remaining comments on unused or maybe unused headers.

**Comment by wdconinc** [comment]
@veprbl is the llvm-cov failure in https://github.com/eic/EICrecon/actions/runs/5495417366/jobs/10015095601?pr=751#step:5:58 a transient issue?

**Comment by wdconinc** [comment]
> transient issue?

Ah, permissions on forks don't include comment posting permission, ref https://docs.github.com/en/actions/security-guides/automatic-token-authentication#permissions-for-the-github_token.

File: src/algorithms/reco/ElectronReconstruction.h (hunk 25 above)
Comment (jdbrice) [reply src/algorithms/reco/ElectronReconstruction.h:10]: resolved with c289e566

File: src/algorithms/reco/ElectronReconstruction.h (hunk 24 above)
Comment (jdbrice) [reply src/algorithms/reco/ElectronReconstruction.h]: resolved with c289e566

File: src/algorithms/reco/ElectronReconstruction.h (hunk 23 above)
Comment (jdbrice) [reply src/algorithms/reco/ElectronReconstruction.h]: resolved with c289e566

File: src/algorithms/reco/ElectronReconstruction.h (hunk 22 above)
Comment (jdbrice) [reply src/algorithms/reco/ElectronReconstruction.h]: resolved with c289e566

File: src/algorithms/reco/ElectronReconstruction.cc (hunk 21 above)
Comment (jdbrice) [reply src/algorithms/reco/ElectronReconstruction.cc]: resolved with c289e566

**Review by wdconinc — APPROVED** [review]

**Comment by wdconinc** [comment]
> Podio supports reference/subset collections

I guess this wasn't actually implemented here??

**Pull request merged**